### DISCUSSION TITLE
単体テストの考え方/使い方 2.4〜3.1.2(くらいまで)

### DISCUSSION THREAD

**Discussion by dameleon** [atman-inc/training#93] [General]
- 2.4 古典学派及びロンドン学派における統合テスト
- 3.1 単体テストの構造
    - 3.1.2くらいまで目標

上記辺りののディスカッション。
不明点、相談点、確認点、勉強になったポイントなどは各自ディスカッションのコメントをお願いします。

**Comment by y-kaneda** [comment]
## 2.4 古典学派およびロンドン学派における統合(integration)テスト

### メモ

#### 統合テストの分類
- ロンドン学派
    - 実際の協力者オブジェクト(collaborator)を使って行うテスト全てを統合テストとみなす
    - 古典学派のスタイルの単体テストはロンドン学派的にはほとんどが統合テストに分類される

- 古典学派
    - 単体テストの定義
        - 単体(unit)と呼ばれる少量のコード: **1単位の振る舞い(a unit of behavior)**　を検証すること
        - **他のテスト・ケースから** 隔離された状態で実行されること
        - 実行時間が短いこと
    - これらの単体テストの定義から外れれば統合テストとなる
    - 例えばケース間でDB依存等があれば隔離されていないので統合テスト
        - 統合ケース間で影響があるのでケースを1つずつ順番に行う必要がある
        - DB等のプロセス外依存にアクセスすることはメモリに展開しているオブジェクト間のやり取りよりはるかに時間がかかる
    - 1つのケース内で複数の振る舞いを検証するケースも統合テスト
        - ただし、そもそもそれぞれのテストケースが長ければ最初から統合テストに分類される
    - 世に言う結合テスト(別チームで作った複数モジュールを組み合わせてテストする)も統合テスト
        - 3要素全部外れることになる

#### E2Eテスト
- 別名：UIテスト、GUIテスト、機能テスト
- E2Eテストは統合テストよりプロセス外依存を多く含む
    - ユーザーのユースケースに沿ったシナリオとなるため
- E2Eテストであっても外部サービスはテスト・ダブルに置き換えたりはする
- E2Eテストの実行は単体・統合テストが全部成功してから

### 感想・疑問点
- 「実行時間が短いこと」の定義は結局のところ何なのか
  - 時間の絶対値によるものではなく、相対値？
  - プロセス外依存等、テスト対象となる該当オブジェクトだけの操作より一般に時間がかかるとされるものを含まないということ？
    - 「少量のコード」の定義とちょっと被る部分がある？

- _これら2つのテスト・ケースは最初から実行時間がかかるものであるため、最初から統合テストに分類されることになります。_ の部分がよく分からなかった
  - 実行時間についてどこかに書いてある？
  - 実行時間を短くするためにケースをまとめたというコンテキストが含まれるから？

- E2Eテストですべて(もしくは、ほぼすべての)プロセス外依存を扱うと書いてあるけど、1つのユースケースに関係するプロセス外依存のこと？
  - E2Eテストの全ケースをまとめて見た際のことを言っている？

## 3.1 単体テストの構造

### メモ

- AAAパターン
  - 準備(Arrange)
  - 実行(Act)
  - 確認(Assert)

- Given-When-Thenパターン
  - Given(前提として) -> Arrange相当
  - When(xxxした時) -> Act相当
  - Then(その結果) -> Assert相当

- テスト駆動開発(TDD)だとプロダクションコードの前にテストコードを書くので、まず先にAssertだけを書き出したりする

- 単体テストでは1つのケースの中に複数の実行フェーズと確認フェーズを持たない
  - 1単位の振る舞いだけを検証する

- 統合テストでは複数のフェーズを含めたほうが効率的な場合がある

### 感想・疑問点
- Arrange 部分、ケースを動かすためのテストデータを作成するコード(fixture)が膨大になりがち。どうにかしたい

- RSpecとかのテスト結果表示(describe 'xxxx' context 'when xxxx' it 'returns false' みたいな)、Given-When-Thenパターンを意識してるのかなと思った。そして [rspec-given](https://github.com/rspec-given/rspec-given) なる gem があった

- 1つの実行フェーズに対して複数の検証フェーズがあるのは統合テスト？
  - どこまで単体という定義に拘るべきか
  - 例えば以下のケースは良いのか悪いのか
    - 画面から集計ジョブをキックするAPIがあったとして、そのusecaseのメソッドを検証する場合
        - 実行フェーズでusecaseのメソッドを実行
        - 確認フェーズ1: ジョブの情報がDBに保存される
        - 確認フェーズ2: SQSのEnqueueがされている

**Reply by y-kaneda** [reply]
結局軽いサマリみたいになってしまった

**Comment by nmurakami0** [comment]
2.4

- 感想
  - 実行時間の記述があるが、必要条件としてはわかるが、十分条件としては必要なのか
    - この条件に引っかかるものは、他の2つのどれかに引っかかる気もする

3.1

- Actフェーズは1行になるはず
  - サービスクラスのrunメソッドがまさにそれに相当
- AAAの各フェーズを空行で区切るか、コメントを付与する
  - 確かにrspec書いているとArrangeフェーズが膨大になるので、コメントつけるのはありかもしれない
- テスト対象システムに対してsutという変数をつける
  - rspecの `subject` がそれに近そう
    - rspec的にはそれにさらに名前をつけるのが推奨されていた気がする
      - https://www.rubydoc.info/gems/rubocop-rspec/1.7.0/RuboCop/Cop/RSpec/NamedSubject

**Comment by xxManatoxx** [comment]
## 古典学派及びロンドン学派における統合テスト

- 今までの単体テストの考え方から統合テストについての説明
- 統合テストとE2Eテストの説明
- 疑問点
  - 古典学派の単体テスト性質説明で必ずしも時間について言及するべきなのかが疑問に思いました。

## 単体テストの構造的解析

- 学び
  - 構造を揃えることによる可読性の向上
  - AAAパターン（Given-When-Thenパターン）は日常の問題解決の流れと同じ手順
- AAAパターン
  - 準備（Arrange）
  - 実行（Act）
  - 確認（Assert）
- 通常は上から実装していく
  - テスト駆動開発の場合確認から書き始める
- Given-When-Thenパターン
  - Given: AAAパターンの準備
  - When: AAAパターンの実行
  - Then: AAAパターンの確認

**Comment by koushisa** [comment]
## 2.4

### 感想・疑問点

- 統合テスト = 結合テスト?
- E2Eテスト
  - UIの実行環境をiframe, mswなどで隔離した場合は統合テストと捉えることができそう
  - プロセス外依存を含まないため
  - 視点がエンドユーザーが、システムなのかという感じか

#### メモ

- 統合(integration)テスト
	- 分類は金田さんのサマりに書いてある
	  - https://github.com/atman-inc/training/discussions/93#discussioncomment-5184567
	- 本書における統合テストの定義は上記の性質を一つでも損っている、ということ
- E2Eテスト
	- 統合テストの一種だがUIなどをエンドユーザーの視点から検証する
	- プロセス外依存をそのまま使う

## 3

### 感想・疑問点

- 読みやすいテストコードを書くためにAAAパターンに準拠せよ
- 同じフェーズが複数あるテストケース
  - 単体テストでは検証したい振る舞いが大きすぎるので分割せよ
  - 統合テストにおいては複数のフェーズをもたせる場合もあるが極力避けるべき

**Comment by k-itachida** [comment]
### 2.4
- 古典学派の単体テスト
  - 1単位の振る舞いを検証すること
  - 実行時間が短いこと
  - 他のテストケースから隔離された状態であること
- これらの性質を１つでも損なっていると統合テストになる
- E2Eテストは統合テストに加えて、ほぼすべてのプロセス外依存を扱う
- 感想
  - テストケースや１単位の振る舞いの具体例がイメージできておらず、テストケースと１単位の振る舞いの意味がごっちゃになってくる。
  - 単体テストは"テストケース = １単位の振る舞い"になるということか。
  - 1つのテストケースで複数の振る舞いを検証する場合は、統合テストになると記述されているが、後々パラメータ化(p.83)のところで１単位の振る舞いで複数のテストケースというパターンも出てくる...

**Comment by myamamoto111** [comment]
## 2.4 統合(integration)テスト
### 感想・疑問点
- テストの呼び方は、時代と関わってきた業務や案件によって異なる気がする
  - 統合テスト
    - 結合テスト、シナリオテスト、機能テスト、etc.
  - E2Eテスト
    - 総合テスト、システムテスト、受け入れテスト、etc.
- どこまでを自動化できて、どこからが手動でしかテストできないのか？
  - やろうと思えば、全て自動化できる？
  - その分、コストがかかる。
  - でも、手動テストもコストがかかる。
- これまでの自分の認識的には
  - 単体テスト
    - 詳細仕様書要件を満たしているか
  - 結合テスト
    - 基本仕様書要件を満たしているか
  - システムテスト
    - 要件定義を満たしているか
